pkg/utils/kubernetes/status.go

```diff
@@ -0,0 +1,329 @@
+// SPDX-FileCopyrightText: 2021 SAP SE or an SAP affiliate company and Gardener contributors.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package kubernetes
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/go-logr/logr"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	apimacherrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	StatusReady      StatusType = "Ready"
+	StatusInProgress StatusType = "InProgress"
+	StatusUnknown    StatusType = "Unknown"
+)
+
+// StatusType defines the value of a Status.
+type StatusType string
+
+// ObjectState represents an object and its status.
+type ObjectState struct {
+	Object *unstructured.Unstructured
+	Status StatusType
+}
+
+// ObjectState represents an arrays of objects and their status.
+type ObjectStates []*ObjectState
+
+// GetObjectsInStatusType returns all unstructured objects in the given status.
+func (o *ObjectStates) GetObjectsInStatusType(status StatusType) []*unstructured.Unstructured {
+	return FilterObjectStateInStatusType(*o, status)
+}
+
+// GetObjectsNotInStatusType returns all unstructured objects not in the given status.
+func (o *ObjectStates) GetObjectsNotInStatusType(status StatusType) []*unstructured.Unstructured {
+	return FilterObjectStateNotInStatusType(*o, status)
+}
+
+// WaitObjectsReady waits for objects to be in ready status and
+// returns an error if all the objects are not ready after the backoff duration.
+func WaitObjectsReady(ctx context.Context, backoff wait.Backoff, log logr.Logger, kubeClient client.Client, objects []*unstructured.Unstructured) error {
```

Comment:
Maybe it would be helpful to have the readiness errors not only as aggregated error but also in the providerStatus.
WDYT?

Reply:
How'd do you see it ? right now the errors show up in `lastError`